File: examples/fireblocks/sign_evm.py
"""Reference: sign an EVM settlement build with Fireblocks (CONTRACT_CALL).

Fireblocks signs AND broadcasts via its own nodes, so you do NOT call client.broadcast for this path.

    pip install hashlock-sdk fireblocks-sdk

Validated live on Fireblocks Sandbox (Sepolia / ETH_TEST5) — the CONTRACT_CALL signs, broadcasts, and the
approve/createSwap lands on-chain. NOTE the Python SDK's shape: peer-type/operation are top-level string
constants (VAULT_ACCOUNT / ONE_TIME_ADDRESS / CONTRACT_CALL), and create_transaction takes `tx_type=`
(not `operation=`). The base URL for sandbox is https://sandbox-api.fireblocks.io (prod: api.fireblocks.io);
the vault account id is a small integer string (e.g. "0").
"""
import os
import time

from fireblocks_sdk import CONTRACT_CALL, ONE_TIME_ADDRESS, VAULT_ACCOUNT, DestinationTransferPeerPath, FireblocksSDK, TransferPeerPath

from hashlock import HashlockClient

client = HashlockClient(api_key=os.environ["HASHLOCK_API_KEY"])
fireblocks = FireblocksSDK(
    private_key=open(os.environ["FIREBLOCKS_SECRET_KEY_PATH"]).read(),
    api_key=os.environ["FIREBLOCKS_API_KEY"],
    api_base_url=os.environ.get("FIREBLOCKS_BASE_URL", "https://sandbox-api.fireblocks.io"),
)

VAULT_ACCOUNT_ID = os.environ["FIREBLOCKS_VAULT_ACCOUNT_ID"]  # numeric string, e.g. "0"
ASSET_ID = "ETH_TEST5"  # Fireblocks asset id for the target chain (Sepolia here)


def sign_via_fireblocks(tx: dict) -> str:
    """Submit one unsigned EVM tx as a Fireblocks contract call; return the on-chain hash."""
    resp = fireblocks.create_transaction(
        tx_type=CONTRACT_CALL,
        asset_id=ASSET_ID,
        source=TransferPeerPath(VAULT_ACCOUNT, VAULT_ACCOUNT_ID),
        destination=DestinationTransferPeerPath(ONE_TIME_ADDRESS, one_time_address={"address": tx["to"]}),
        amount=str(int(tx.get("value") or "0")),  # wei; "0" for approve / createSwap
        extra_parameters={"contractCallData": tx["data"]},
        note="Hashlock HTLC settlement",
    )
    tx_id = resp["id"]
    while True:
        t = fireblocks.get_transaction_by_id(tx_id)
        if t.get("txHash"):
            return t["txHash"]
        if t["status"] in ("FAILED", "BLOCKED", "CANCELLED", "REJECTED"):
            raise RuntimeError(f"Fireblocks tx {tx_id} {t['status']}: {t.get('subStatus')}")
        time.sleep(3)


def fund_evm_leg(swap_id: str, leg: str) -> None:
    build = client.build_fund(swap_id, leg)  # {'sign': 'evm-tx', 'txs': [approve, createSwap], ...}
    for tx in build["txs"]:
        print("fireblocks broadcast", sign_via_fireblocks(tx))

File: examples/fireblocks/sign_evm.ts
/**
 * Reference: sign an EVM settlement build with Fireblocks (CONTRACT_CALL). Fireblocks signs AND broadcasts
 * via its own nodes, so you do NOT call client.broadcast for this path — you poll Fireblocks for the hash.
 *
 *   npm i @hashlock-tech/sdk fireblocks-sdk
 *
 * The CONTRACT_CALL flow is validated live on Fireblocks Sandbox (Sepolia); the `PeerType` /
 * `TransactionOperation` / `TransactionStatus` enums below match the current `fireblocks-sdk`. Set your
 * own vault account id + asset id. (Python uses different symbols — see sign_evm.py / the README.)
 */
import { HashlockClient, type EvmBuild } from '@hashlock-tech/sdk';
import { FireblocksSDK, PeerType, TransactionOperation, TransactionStatus } from 'fireblocks-sdk';

const client = new HashlockClient({ apiKey: process.env.HASHLOCK_API_KEY! });
const fireblocks = new FireblocksSDK(process.env.FIREBLOCKS_API_SECRET!, process.env.FIREBLOCKS_API_KEY!);

const VAULT_ACCOUNT_ID = process.env.FIREBLOCKS_VAULT_ID!; // your vault account
const ASSET_ID = 'ETH_TEST5'; // Fireblocks asset id for the target chain (Sepolia here)

/** Submit one unsigned EVM tx as a Fireblocks contract call and wait for the on-chain hash. */
async function signViaFireblocks(tx: { to: string; data: string; value?: string }): Promise<string> {
  const { id } = await fireblocks.createTransaction({
    operation: TransactionOperation.CONTRACT_CALL,
    assetId: ASSET_ID,
    source: { type: PeerType.VAULT_ACCOUNT, id: VAULT_ACCOUNT_ID },
    destination: { type: PeerType.ONE_TIME_ADDRESS, oneTimeAddress: { address: tx.to } },
    amount: tx.value ? String(BigInt(tx.value)) : '0', // wei; '0' for ERC-20 approve / createSwap
    extraParameters: { contractCallData: tx.data },
    note: 'Hashlock HTLC settlement',
  });

  // Return only once the tx is mined (CONFIRMING) or fully confirmed (COMPLETED) — NOT on the mere
  // presence of a txHash — so the dependent createSwap isn't submitted before this approve lands
  // (createSwap's transferFrom needs the approve's allowance on-chain).
  for (;;) {
    const t = await fireblocks.getTransactionById(id);
    if ((t.status === TransactionStatus.CONFIRMING || t.status === TransactionStatus.COMPLETED) && t.txHash) return t.txHash;
    if ([TransactionStatus.FAILED, TransactionStatus.BLOCKED, TransactionStatus.CANCELLED, TransactionStatus.REJECTED].includes(t.status)) {
      throw new Error(`Fireblocks tx ${id} ${t.status}: ${t.subStatus}`);
    }
    await new Promise((r) => setTimeout(r, 3000));
  }
}

export async function fundEvmLeg(swapId: string, leg: 'a' | 'b') {
  const build = (await client.buildFund(swapId, leg)) as EvmBuild;
  // EVM fund = [approve, createSwap]; submit in order, waiting for each (approve must confirm before createSwap).
  for (const tx of build.txs) {
    const hash = await signViaFireblocks(tx);
    console.log('fireblocks broadcast', hash);
  }
}
